src/transformStateWithClones.js

```diff
@@ -7,7 +7,37 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    let newState = { ...currentState };
+
+    const { type, extraData, keysToRemove } = action;
+
+    switch (type) {
+      case 'addProperties':
+        for (const [key, value] of Object.entries(extraData)) {
+          newState[key] = value;
+        }
+        break;
+
+      case 'removeProperties':
+        for (const key of keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        newState = {};
+        break;
+    }
```

Comment:
[CODE STYLE]: The `switch` statement is missing a `default` case. It's important to have a `default` case to handle any unexpected `type` values. This can be a simple case that either throws an error or continues without modifying the state.